File: Arrays/3sumclosest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int length = nums.size();
        int left, right;
        int result = nums[0]+nums[1]+nums[2];
        int cur = abs(result-target);
        for (int i=0; i<length -2; i++){
            int temp = nums[i]-target;
            left = i+1;
            right = length -1;
            while(left<right){
                int temp1 = temp+nums[left]+nums[right];
                if (abs(temp1)<cur){
                    result = nums[i]+nums[left]+nums[right];
                    if(temp1==0){
                       return result;
                    }
                    cur = abs(result-target);
                }
                if (temp1 >0)
                    right--;
                if (temp1 < 0)
                    left++;
                
            }
        }
        return result;
    }
};

File: Arrays/3sumclosest.py
class Solution:
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        result = []
        if len(nums)>0:
            for i in range(len(nums)-2):
                number = nums[i]
                left = i+1
                right = len(nums)-1
                while(left<right):
                    total_sum = nums[left]+nums[right]+number
                    if total_sum == target:
                        return total_sum
                    elif total_sum > target:
                        right = right -1
                    else:
                        left = left + 1
                    result.append(total_sum)
            return min(result, key = lambda x:abs(x-target) )
                    
            
        
